ENH: add metadata necessary to download from previous idc-versions

vkt1414 opened the pull request:
Pulls the strictly necessary metadata to allow downloading data from previous idc versions

solves https://github.com/ImagingDataCommons/idc-index/issues/100

fedorov commented:
Wow, this looks really great from the description - I will review tomorrow!

How about renaming the query and the result int prior_versions_index? Or some other ..._index.

It is a key step towards solving #100, but to actually solve it, we would need to integrate handling of this additional table in `idc-index`.

fedorov commented on assets/prior_versions_index.sql:
```diff
@@ -0,0 +1,77 @@
+  MIN(idc_version) AS min_idc_version,
+  MAX(idc_version) AS max_idc_version
+FROM all_versions
+where gcs_bucket not in ('idc-open-idc')
```
Can you add inline comment why this bucket is excluded?

vkt1414 replied:
I do not have any other reason other than trying to match Bill's table. We can allow this bucket as well.

fedorov commented on assets/prior_versions_index.sql:
```diff
@@ -0,0 +1,77 @@
+  crdc_series_uuid,
+  series_size_MB,
+  CASE
+  WHEN gcs_bucket='public-datasets-idc' THEN CONCAT('s3://','idc-open-data/',crdc_series_uuid, '/*')
```
```suggestion
  # map GCS bucket to AWS bucket, since for idc-index we prefer AWS
  # if new buckets are included in IDC, this will need to be updated!
  WHEN gcs_bucket='public-datasets-idc' THEN CONCAT('s3://','idc-open-data/',crdc_series_uuid, '/*')
```

fedorov commented on assets/prior_versions_index.sql:
```diff
@@ -0,0 +1,77 @@
+  GROUP BY
+  1,2,3,4,5,6,7,8
+
+""", version, version, latest_idc_version),
```
```suggestion
  """, 
  version, version, latest_idc_version),
```

fedorov replied:
This is for better readability, since otherwise BQ console marks everything that follows `"""` as a comment, and that confused me a lot when I first looked at it.

fedorov commented on assets/prior_versions_index.sql:
```diff
@@ -0,0 +1,77 @@
+--series_instance_uid not in (select distinct seriesInstanceUID from bigquery-public-data.idc_current.dicom_all))
+GROUP BY
+ 1,2,3,4,5,6,7,8
+""", union_all_query
```
```suggestion
  """, 
  union_all_query
```

fedorov commented on assets/prior_versions_index.sql:
```diff
@@ -0,0 +1,77 @@
+-- Step 1: Declare variables
+DECLARE idc_versions ARRAY<INT64>;
+DECLARE latest_idc_version INT64;
```
```suggestion
DECLARE latest_idc_version INT64 DEFAULT 18;
```

fedorov commented on assets/prior_versions_index.sql:
```diff
@@ -0,0 +1,77 @@
+
+-- Step 2: Get all idc_versions
+SET idc_versions = (
+  SELECT [1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18]
```
```suggestion
  SELECT GENERATE_ARRAY(1, latest_idc_version)
  -- SELECT [1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18]
```

vkt1414 replied:
this is only temporary and will be updated dynamically with commented out query once Bill updates the `bigquery-public-data.idc_current.version_metadata` metadata table

fedorov commented on assets/prior_versions_index.sql:
```diff
@@ -0,0 +1,77 @@
+  --`bigquery-public-data.idc_current.version_metadata`
+);
+
+SET latest_idc_version = (
+  SELECT 18
+  --SELECT max(idc_version)
+  --FROM
+  --`bigquery-public-data.idc_current.version_metadata`
+);
```
```suggestion
```

fedorov reviewed (changes requested):
I just have some small suggestions.

vkt1414 commented:
@fedorov I just want to make sure you want to update the `latest_idc_version`  variable yourself.

https://github.com/ImagingDataCommons/idc-index-data/pull/32/files#diff-1b8acf7ccf6a7bc784d04a22bb82263a847dacb31532f44d6c046ada6540e24eR3

fedorov commented:
> @fedorov I just want to make sure you want to update the latest_idc_version variable yourself.

As opposed to running daily github action and doing this automatically? If that's the question, then yes - I don't see a need in automating that.

vkt1414 commented:
> > @fedorov I just want to make sure you want to update the latest_idc_version variable yourself.
> 
> As opposed to running daily github action and doing this automatically? If that's the question, then yes - I don't see a need in automating that.

Not a github action. Initially I had the following code to get the latest idc version dynamically. I hardcoded to v18 for now as Bill needs to update the table -`bigquery-public-data.idc_current.version_metadata`. But you removed it in your suggestion and I wanted to know if you'll change when switching idc version. And it looks you are. So I'm ok with that. 

SET latest_idc_version = (
  SELECT 18
  --SELECT max(idc_version)
  --FROM 
  --`bigquery-public-data.idc_current.version_metadata`
);

fedorov commented:
I am pretty sure you had it hardcoded as well - if I remember correctly, I just replaced "SELECT 18" with a variable initialization. The other lines were commented out. I don't have any problem keeping the commented out lines in the query, if you prefer. We can uncomment whenever that table is fixed. I don't recall Bill mentioning he is working on that, so I don't know about the timeline.

vkt1414 commented:
> I am pretty sure you had it hardcoded as well - if I remember correctly, I just replaced "SELECT 18" with a variable initialization. The other lines were commented out. I don't have any problem keeping the commented out lines in the query, if you prefer. We can uncomment whenever that table is fixed. I don't recall Bill mentioning he is working on that, so I don't know about the timeline.

I added the commented out code back. Once Bill updates the table, we can revisit the query
